src/Features/CSharp/Portable/MakeStructFieldsWritable/CSharpMakeStructFieldsWritableDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,120 @@
+            => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(compilationStartContext
+                => SymbolAnalyzer.CreateAndRegisterActions(compilationStartContext));
+        }
+
+        private sealed class SymbolAnalyzer
+        {
+            private readonly INamedTypeSymbol _namedTypeSymbol;
+            private bool _hasTypeInstanceAssigment;
+
+            private SymbolAnalyzer(INamedTypeSymbol namedTypeSymbol)
+            {
+                _namedTypeSymbol = namedTypeSymbol;
+            }
+
+            public static void CreateAndRegisterActions(CompilationStartAnalysisContext compilationStartContext)
+            {
+                compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+                {
+                    // We report diagnostic only if these requirements are met:
+                    // 1. The type is struct
+                    // 2. Struct contains at least one 'readonly' field
+                    // 3. Struct contains assignment to 'this' outside the scope of constructor
+                    var namedTypeSymbol = (INamedTypeSymbol)symbolStartContext.Symbol;
+
+                    // We are only interested in struct declarations
+                    if (namedTypeSymbol.TypeKind != TypeKind.Struct)
+                    {
+                        return;
+                    }
+
+                    //We check if struct contains any 'readonly' fields
+                    if (!HasReadonlyField(namedTypeSymbol))
+                    {
+                        return;
+                    }
+
+                    var symbolAnalyzer = new SymbolAnalyzer(namedTypeSymbol);
+                    symbolAnalyzer.RegisterActions(symbolStartContext);
+                }, SymbolKind.NamedType);
+            }
+
+            private static bool HasReadonlyField(INamedTypeSymbol namedTypeSymbol)
+            {
+                return namedTypeSymbol
+                    .GetMembers()
+                    .OfType<IFieldSymbol>()
+                    .Where(field => field.AssociatedSymbol == null)
+                    .Any(field => field.IsReadOnly);
+            }
+
+            private void RegisterActions(SymbolStartAnalysisContext symbolStartContext)
+            {
+                symbolStartContext.RegisterOperationBlockStartAction(blockAction =>
+                {
+                    var isConstructor = false;
+                    blockAction.RegisterOperationAction(
+                        _ => isConstructor = true, OperationKind.ConstructorBody);
+                    blockAction.RegisterOperationAction(
+                        operationAction => AnalyzeAssignment(operationAction, isConstructor), OperationKind.SimpleAssignment);
```

Comment:
You should not rely on the order of IOperation callbacks here. For example, the analyzer driver is free to call back all `OperationKind.SimpleAssignment` nodes before `OperationKind.ConstructorBody`. An easier way here is to just use `var isConstructor = blockAction.OwningSymbol is IMethodSymbol method && method.MethodKind == MethodKind.Constructor` and avoid registering operation callback for `OperationKind.ConstructorBody`

Reply:
Thanks for your feedback.
Resolved in [https://github.com/dotnet/roslyn/pull/32063/commits/e845e1daf674a70d3f6b82f6ec3c68e365f2282e](https://github.com/dotnet/roslyn/pull/32063/commits/e845e1daf674a70d3f6b82f6ec3c68e365f2282e)